File: 프로그래머스/2/42842. 카펫/카펫.cpp
#include <vector>
#include <cmath>

using namespace std;

vector<int> solution(int brown, int yellow) {
    int total = brown + yellow; // 전체 카펫 크기

    for (int height = 3; height <= sqrt(total); ++height) {
        if (total % height == 0) { // 전체 면적이 나누어 떨어지는 경우
            int width = total / height;
            if ((width - 2) * (height - 2) == yellow) { // 노란색 크기 확인
                return {width, height};
            }
        }
    }
    return {}; // 조건을 만족하는 경우가 없을 때 (문제에서는 주어지지 않음)
}

File: 프로그래머스/2/42842. 카펫/카펫.cs
using System;
using System.Collections.Generic;

public class Solution 
{
    bool[] visited;
    HashSet<(int w, int h)> hash = new HashSet<(int, int)>();
    public int[] solution(int brown, int yellow) 
    {
        int[] result = new int[2];
        visited = new bool[brown];
        DFS(brown, yellow);
        foreach(var i in hash)
        {
            if(i.w >= i.h)
            {
                if((i.w - 2) * (i.h - 2) == yellow)
                {
                    Console.WriteLine(i);
                    result[0] = i.w;
                    result[1] = i.h;
                }
            }
        }
        return result;
    }
    
    void DFS(int b, int y)
    {
        int sum = b + y;
        for(int i=b; i>=1; i--)
        {
            if(sum % i == 0)
            {
                hash.Add((i, sum / i)); 
            }
        }
        
        
    }
}